packages/firestore/lib/modular/snapshot.d.ts

```diff
@@ -1,14 +1,195 @@
-import { FirebaseFirestoreTypes } from '../index';
-
-import DocumentReference = FirebaseFirestoreTypes.DocumentReference;
-import DocumentSnapshot = FirebaseFirestoreTypes.DocumentSnapshot;
-import SnapshotListenOptions = FirebaseFirestoreTypes.SnapshotListenOptions;
-import QuerySnapshot = FirebaseFirestoreTypes.QuerySnapshot;
-import Query = FirebaseFirestoreTypes.Query;
+import { DocumentReference, Query, DocumentData, SnapshotListenOptions } from './index';
 
 export type Unsubscribe = () => void;
 export type FirestoreError = Error;
 
+/**
+ * A DocumentSnapshot contains data read from a document in your Firestore database. The data can be extracted with
+ * .`data()` or `.get(:field)` to get a specific field.
+ *
+ * For a DocumentSnapshot that points to a non-existing document, any data access will return 'undefined'.
+ * You can use the `exists()` method to explicitly verify a document's existence.
+ */
+export interface DocumentSnapshot<
+  AppModelType = DocumentData,
+  DbModelType extends DocumentData = DocumentData,
+> {
+  /**
+   * Method of the `DocumentSnapshot` that signals whether or not the data exists. True if the document exists.
+   */
+  exists(): this is QueryDocumentSnapshot<AppModelType, DbModelType>;
+
+  /**
+   * Property of the `DocumentSnapshot` that provides the document's ID.
+   */
+  id: string;
+
+  /**
+   * Metadata about the `DocumentSnapshot`, including information about its source and local modifications.
+   */
+  metadata: SnapshotMetadata;
+
+  /**
+   * The `DocumentReference` for the document included in the `DocumentSnapshot`.
+   */
+  ref: DocumentReference<AppModelType, DbModelType>;
+
+  /**
+   * Retrieves all fields in the document as an Object. Returns 'undefined' if the document doesn't exist.
+   *
+   * #### Example
+   *
+   * ```js
+   * const user = await firebase.firestore().doc('users/alovelace').get();
+   *
+   * console.log('User', user.data());
+   * ```
+   */
+  data(): AppModelType | undefined;
+
+  /**
+   * Retrieves the field specified by fieldPath. Returns undefined if the document or field doesn't exist.
+   *
+   * #### Example
+   *
+   * ```js
+   * const user = await firebase.firestore().doc('users/alovelace').get();
+   *
+   * console.log('Address ZIP Code', user.get('address.zip'));
+   * ```
+   *
+   * @param fieldPath The path (e.g. 'foo' or 'foo.bar') to a specific field.
+   */
+  get<fieldType extends DbModelType[keyof DbModelType]>(
+    fieldPath: keyof DbModelType | string | FieldPath,
+  ): fieldType;
```

Comment:
Thx for the nudge @mikehardy, I would suggest this:

```suggestion
  get<K extends keyof DbModelType | FieldPath>(
    fieldPath: K,
  ): K extends keyof DbModelType ? DbModelType[K] : any;
```

If the caller passes a string literal, they get back the type of the field with that key. It seems unnecessary to include `string`: DocumentData is the default for DbModelType, and DocumentData is Record<string, any>. So that will allow K to be any string and return any. Meanwhile if a specific type is used for DbModelType, that will restrict K to known keys of that type or FieldPath objects.

Passing a FieldPath can access nested paths, and there's no way for the types to resolve what type is something like `get(new FieldPath(["my", "nested", "property"]))`. FieldPath could also be FieldPath.documentId() in which case the return type is a string. So `any` is the simplest thing to return there. One could make FieldPath generic and retain type info about the full range of keys, and then you could resolve the nested properties with a recursive type helper, but I imagine the complexity isn't worth it.